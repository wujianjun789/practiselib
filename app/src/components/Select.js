/**
 * Created by a on 2017/7/7.
 */
import React, { Component } from 'react'
import Immutable from 'immutable';

export default class Select extends Component {
    constructor(props) {
        super(props)
        this.onChange = this.onChange.bind(this);
    }

    onChange(event) {
        this.props.onChange && this.props.onChange(event.target.selectedIndex);
    }

    render() {
        const {className='', id='', valueKey="value", titleKey="value", data=Immutable.fromJS({
                list: [],
                value: '',
                placeholder: ""
            })} = this.props;

        return <select className={ "select " + className + " form-control" } id={ id } placeholder={ data.get('placeholder') } value={ data.get("value") } onChange={ this.onChange }>
                 { data.get('list').map((option, index) => {
                       let value = option.get(valueKey);
                       let title = option.get(titleKey);
                       return <option key={ index } value={ value }>
                                { title }
                              </option>
                   }) }
               </select>
    }
}