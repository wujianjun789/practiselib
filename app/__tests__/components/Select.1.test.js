import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import Select from '../../src/components/Select.1';

describe('<Select.1 />>', () => {
    let options = [
        {title: 'value01', value: 'value01'},
        {title: 'value02', value: 'value02'},
        {title: 'value03', value: 'value03'}
    ];
    let className="select-primary";
    let value = 'value02';

    it('render', () => {
        const cmp = shallow(<Select />);
        let select = cmp.find('select');
        expect(select.hasClass('select')).toBeTruthy();
        expect(select.prop('value')).toBe('');
        let opts = cmp.find('option');
        expect(opts.length).toBe(0);
    });

    it('render with className="select-primary"', () => {
        const cmp = shallow(<Select className={className} options={options} value={value}/>);
        let select = cmp.find('select');
        expect(select.hasClass('select select-primary')).toBeTruthy();
        expect(select.props().value).toBe(value);

        let opts = cmp.find('option');
        expect(opts.length).toBe(options.length);
    });

    it('render with titleField="title" valueField="value"', () => {
        let titleField = 'title';
        let valueField = 'value';

        const cmp = shallow(<Select titleField="title" valueField="value" options={options} value="value03"/>);
        let opts = cmp.find('option');
        expect(opts.at(0).prop('value')).toBe(options[0].value);
        expect(opts.at(0).text()).toBe(options[0].title);
    });

    it('onChange', () => {
        let onChange = jest.fn();
        const cmp = shallow(<Select onChange={onChange}/>);

        cmp.find('select').simulate('change');
        expect(onChange).toHaveBeenCalledTimes(1);
    });

    it('snapshot', () => {
        const cmp = renderer.create(<Select className={className} titleField="title" valueField='value' options={options} value={value} />);
        let tree = cmp.toJSON();
        expect(tree).toMatchSnapshot();
    })

})