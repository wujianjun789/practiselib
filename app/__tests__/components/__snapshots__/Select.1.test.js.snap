exports[`<Select.1 />> snapshot 1`] = `
<select
  className="form-control select select-primary"
  disabled={undefined}
  id=""
  onChange={undefined}
  value="value02">
  <option
    value="value01">
    value01
  </option>
  <option
    value="value02">
    value02
  </option>
  <option
    value="value03">
    value03
  </option>
</select>
`;
